doc/developers/contributing.rst

```diff
@@ -603,6 +603,24 @@ Finally, follow the formatting rules below to make it consistently good:
     SelectKBest : Select features based on the k highest scores.
     SelectFpr : Select features based on a false positive rate test.
 
+* When documenting the parameters and attributes, have the following in mind:
+
+    1. Do not use optional. Use `default=`. `str {'a', 'b'} or float, default=1.0`
+    2. Python basic types. (`bool` instead of `boolean`)
+    3. When defining 1-D shape, use parenthesis. `array-like shape=(n_samples,), None, default=None`.
+    4. When defining 2-D shape, use parenthesis. `array-like shape=(n_samples, n_features)`.
+    5. str with multiple options: `input: {'log', 'squared', 'multinomial'}`
+    6. Only use `or` for separating types. `float, int or None, default=None`
+    7. Only use comma to separate types. Information such as `shape` or `string`
+    options are defined together. Here is an extreme example:
+
+    ```
+    {'a', 'b'}, float, int, array-like shape=(n_samples,) or None, default=None
```

Comment:
If `shape=` is a standard then I am fine with it. But I would rather find the following easier to parse visually.

```python
"""
array-like of shape (n_samples,)
{array-like, sparse matrix or dataframe} of shape (n_samples, n_features)
"""
```

Reply:
I am okay with `of shape`. I would go with:

```
{array-like, sparse matrix, dataframe} of shape (n_samples, n_features)
```

(There is no "or" before "dataframe")